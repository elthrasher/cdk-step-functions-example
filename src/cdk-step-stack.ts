import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { RetentionDays } from 'aws-cdk-lib/aws-logs';
import { Chain, Choice, Condition, DefinitionBody, Fail, StateMachine, TaskInput } from 'aws-cdk-lib/aws-stepfunctions';
import { LambdaInvoke } from 'aws-cdk-lib/aws-stepfunctions-tasks';
import { Stack, StackProps } from 'aws-cdk-lib/core';
import { Construct } from 'constructs';

const lambdaPath = `${__dirname}/lambda`;

export class CdkStepStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const lambdaProps = {
      logRetention: RetentionDays.ONE_DAY,
      runtime: Runtime.NODEJS_LATEST,
    };

    const assignCaseLambda = new NodejsFunction(this, 'assignCaseFunction', {
      ...lambdaProps,
      entry: `${lambdaPath}/assign-case.ts`,
    });

    const closeCaseLambda = new NodejsFunction(this, 'closeCaseFunction', {
      ...lambdaProps,
      entry: `${lambdaPath}/close-case.ts`,
    });

    const escalateCaseLambda = new NodejsFunction(this, 'escalateCaseFunction', {
      ...lambdaProps,
      entry: `${lambdaPath}/escalate-case.ts`,
    });

    const openCaseLambda = new NodejsFunction(this, 'openCaseFunction', {
      ...lambdaProps,
      entry: `${lambdaPath}/open-case.ts`,
    });

    const workOnCaseLambda = new NodejsFunction(this, 'workOnCaseFunction', {
      ...lambdaProps,
      entry: `${lambdaPath}/work-on-case.ts`,
    });

    const jsonataprops = {
      assign: {
        message: '{% $states.result.Payload.message %}',
        status: '{% $states.result.Payload.status %}',
      },
      payload: TaskInput.fromObject({
        caseId: '{% $caseId %}',
        message: '{% $message %}',
        status: '{% $status %}',
      }),
    };

    const assignCase = LambdaInvoke.jsonata(this, 'Assign Case', {
      ...jsonataprops,
      lambdaFunction: assignCaseLambda,
    });

    const closeCase = LambdaInvoke.jsonata(this, 'Close Case', { ...jsonataprops, lambdaFunction: closeCaseLambda });

    const escalateCase = LambdaInvoke.jsonata(this, 'Escalate Case', {
      ...jsonataprops,
      lambdaFunction: escalateCaseLambda,
    });

    const openCase = LambdaInvoke.jsonata(this, 'Open Case', {
      assign: {
        caseId: '{% $states.input.inputCaseID %}',
        message: '{% $states.result.Payload.message %}',
        status: '{% $states.result.Payload.status %}',
      },
      lambdaFunction: openCaseLambda,
    });

    const workOnCase = LambdaInvoke.jsonata(this, 'Work On Case', {
      ...jsonataprops,
      lambdaFunction: workOnCaseLambda,
    });

    const jobFailed = Fail.jsonata(this, 'Fail', {
      cause: 'Engage Tier 2 Support',
    });

    const isComplete = Choice.jsonata(this, 'Is Case Resolved');

    const chain = Chain.start(openCase)
      .next(assignCase)
      .next(workOnCase)
      .next(
        isComplete
          .when(Condition.jsonata('{% $status = "closed" %}'), closeCase)
          .when(Condition.jsonata('{% $status = "escalated" %}'), escalateCase.next(jobFailed)),
      );

    new StateMachine(this, 'StateMachine', {
      definitionBody: DefinitionBody.fromChainable(chain),
    });
  }
}
