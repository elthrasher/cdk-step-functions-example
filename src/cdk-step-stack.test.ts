import { Template } from 'aws-cdk-lib/assertions';
import { App } from 'aws-cdk-lib';

import { CdkStepStack } from './cdk-step-stack';

test('Step Functions for Cases', () => {
  const app = new App();
  const stack = new CdkStepStack(app, 'MyTestStack');
  const template = Template.fromStack(stack);

  template.resourceCountIs('AWS::IAM::Policy', 2);
  template.resourceCountIs('AWS::IAM::Role', 7);
  template.resourceCountIs('AWS::Lambda::Function', 6);
  template.resourceCountIs('AWS::StepFunctions::StateMachine', 1);

  expect(template.toJSON()).toMatchSnapshot();
});
