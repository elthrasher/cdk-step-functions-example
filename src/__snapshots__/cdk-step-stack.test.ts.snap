// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step Functions for Cases 1`] = `
{
  "Mappings": {
    "LatestNodeRuntimeMap": {
      "af-south-1": {
        "value": "nodejs20.x",
      },
      "ap-east-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-2": {
        "value": "nodejs20.x",
      },
      "ap-northeast-3": {
        "value": "nodejs20.x",
      },
      "ap-south-1": {
        "value": "nodejs20.x",
      },
      "ap-south-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-1": {
        "value": "nodejs20.x",
      },
      "ap-southeast-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-3": {
        "value": "nodejs20.x",
      },
      "ap-southeast-4": {
        "value": "nodejs20.x",
      },
      "ap-southeast-5": {
        "value": "nodejs20.x",
      },
      "ap-southeast-7": {
        "value": "nodejs20.x",
      },
      "ca-central-1": {
        "value": "nodejs20.x",
      },
      "ca-west-1": {
        "value": "nodejs20.x",
      },
      "cn-north-1": {
        "value": "nodejs20.x",
      },
      "cn-northwest-1": {
        "value": "nodejs20.x",
      },
      "eu-central-1": {
        "value": "nodejs20.x",
      },
      "eu-central-2": {
        "value": "nodejs20.x",
      },
      "eu-isoe-west-1": {
        "value": "nodejs18.x",
      },
      "eu-north-1": {
        "value": "nodejs20.x",
      },
      "eu-south-1": {
        "value": "nodejs20.x",
      },
      "eu-south-2": {
        "value": "nodejs20.x",
      },
      "eu-west-1": {
        "value": "nodejs20.x",
      },
      "eu-west-2": {
        "value": "nodejs20.x",
      },
      "eu-west-3": {
        "value": "nodejs20.x",
      },
      "il-central-1": {
        "value": "nodejs20.x",
      },
      "me-central-1": {
        "value": "nodejs20.x",
      },
      "me-south-1": {
        "value": "nodejs20.x",
      },
      "mx-central-1": {
        "value": "nodejs20.x",
      },
      "sa-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-2": {
        "value": "nodejs20.x",
      },
      "us-gov-east-1": {
        "value": "nodejs20.x",
      },
      "us-gov-west-1": {
        "value": "nodejs20.x",
      },
      "us-iso-east-1": {
        "value": "nodejs18.x",
      },
      "us-iso-west-1": {
        "value": "nodejs18.x",
      },
      "us-isob-east-1": {
        "value": "nodejs18.x",
      },
      "us-west-1": {
        "value": "nodejs20.x",
      },
      "us-west-2": {
        "value": "nodejs20.x",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "2819175352ad1ce0dae768e83fc328fb70fb5f10b4a8ff0ccbcb791f02b0716d.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "LatestNodeRuntimeMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "StateMachine2E01A3A5": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "StateMachineRoleDefaultPolicyDF1E6607",
        "StateMachineRoleB840431D",
      ],
      "Properties": {
        "DefinitionString": {
          "Fn::Join": [
            "",
            [
              "{"StartAt":"Open Case","States":{"Open Case":{"QueryLanguage":"JSONata","Next":"Assign Case","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","Arguments":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "openCaseFunction64DF09DE",
                  "Arn",
                ],
              },
              "","Payload":"{% $states.input %}"},"Assign":{"caseId":"{% $states.input.inputCaseID %}","message":"{% $states.result.Payload.message %}","status":"{% $states.result.Payload.status %}"},"Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke"},"Assign Case":{"QueryLanguage":"JSONata","Next":"Work On Case","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","Arguments":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "assignCaseFunction59553755",
                  "Arn",
                ],
              },
              "","Payload":{"caseId":"{% $caseId %}","message":"{% $message %}","status":"{% $status %}"}},"Assign":{"message":"{% $states.result.Payload.message %}","status":"{% $states.result.Payload.status %}"},"Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke"},"Work On Case":{"QueryLanguage":"JSONata","Next":"Is Case Resolved","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","Arguments":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "workOnCaseFunction65F13148",
                  "Arn",
                ],
              },
              "","Payload":{"caseId":"{% $caseId %}","message":"{% $message %}","status":"{% $status %}"}},"Assign":{"message":"{% $states.result.Payload.message %}","status":"{% $states.result.Payload.status %}"},"Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke"},"Is Case Resolved":{"Type":"Choice","QueryLanguage":"JSONata","Choices":[{"Condition":"{% $status = \\"closed\\" %}","Next":"Close Case"},{"Condition":"{% $status = \\"escalated\\" %}","Next":"Escalate Case"}]},"Close Case":{"QueryLanguage":"JSONata","End":true,"Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","Arguments":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "closeCaseFunctionF5FEF3D0",
                  "Arn",
                ],
              },
              "","Payload":{"caseId":"{% $caseId %}","message":"{% $message %}","status":"{% $status %}"}},"Assign":{"message":"{% $states.result.Payload.message %}","status":"{% $states.result.Payload.status %}"},"Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke"},"Escalate Case":{"QueryLanguage":"JSONata","Next":"Fail","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","Arguments":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "escalateCaseFunction2DE16F3F",
                  "Arn",
                ],
              },
              "","Payload":{"caseId":"{% $caseId %}","message":"{% $message %}","status":"{% $status %}"}},"Assign":{"message":"{% $states.result.Payload.message %}","status":"{% $states.result.Payload.status %}"},"Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke"},"Fail":{"Type":"Fail","QueryLanguage":"JSONata","Cause":"Engage Tier 2 Support"}}}",
            ],
          ],
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "StateMachineRoleB840431D",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
      "UpdateReplacePolicy": "Delete",
    },
    "StateMachineRoleB840431D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "states.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "StateMachineRoleDefaultPolicyDF1E6607": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "openCaseFunction64DF09DE",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "openCaseFunction64DF09DE",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "assignCaseFunction59553755",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "assignCaseFunction59553755",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "workOnCaseFunction65F13148",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "workOnCaseFunction65F13148",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "closeCaseFunctionF5FEF3D0",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "closeCaseFunctionF5FEF3D0",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "escalateCaseFunction2DE16F3F",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "escalateCaseFunction2DE16F3F",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "StateMachineRoleDefaultPolicyDF1E6607",
        "Roles": [
          {
            "Ref": "StateMachineRoleB840431D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "assignCaseFunction59553755": {
      "DependsOn": [
        "assignCaseFunctionServiceRole56B929C6",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "928051b30f8d6e4b7e4623aecf47c2ca3487162efd30e954ee297412d8cc6cad.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "assignCaseFunctionServiceRole56B929C6",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "assignCaseFunctionLogRetention3DDBAA6B": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "assignCaseFunction59553755",
              },
            ],
          ],
        },
        "RetentionInDays": 1,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "assignCaseFunctionServiceRole56B929C6": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "closeCaseFunctionF5FEF3D0": {
      "DependsOn": [
        "closeCaseFunctionServiceRole0F6F67D6",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "a7c96f179720806127cf86b1a508ad996cead5e4669c7799e86c17551409477b.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "closeCaseFunctionServiceRole0F6F67D6",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "closeCaseFunctionLogRetentionCC1C54AA": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "closeCaseFunctionF5FEF3D0",
              },
            ],
          ],
        },
        "RetentionInDays": 1,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "closeCaseFunctionServiceRole0F6F67D6": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "escalateCaseFunction2DE16F3F": {
      "DependsOn": [
        "escalateCaseFunctionServiceRole1BAB0E40",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "a61b4df1211250b0ed4da7a749b9aaa0cbe49432a08fdcdec0a6c274a1faea10.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "escalateCaseFunctionServiceRole1BAB0E40",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "escalateCaseFunctionLogRetentionC5151EC6": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "escalateCaseFunction2DE16F3F",
              },
            ],
          ],
        },
        "RetentionInDays": 1,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "escalateCaseFunctionServiceRole1BAB0E40": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "openCaseFunction64DF09DE": {
      "DependsOn": [
        "openCaseFunctionServiceRole5FA58BBF",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "a67699fb66561d8f5ab354597be1bfb6056d5f128306d2f88c5bd5207600860d.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "openCaseFunctionServiceRole5FA58BBF",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "openCaseFunctionLogRetentionD74F2506": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "openCaseFunction64DF09DE",
              },
            ],
          ],
        },
        "RetentionInDays": 1,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "openCaseFunctionServiceRole5FA58BBF": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "workOnCaseFunction65F13148": {
      "DependsOn": [
        "workOnCaseFunctionServiceRole271B4440",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "cf6dcd17da9fa20edb0f085b4e99110e8a9043737bf0da2561ddbb14a021e0f9.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "workOnCaseFunctionServiceRole271B4440",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "workOnCaseFunctionLogRetentionC3D7D021": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "workOnCaseFunction65F13148",
              },
            ],
          ],
        },
        "RetentionInDays": 1,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "workOnCaseFunctionServiceRole271B4440": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
